### PR TITLE
Rewrite @ScalaSignature when shading

### PR DESCRIPTION
**Pull request by jeroentervoorde**
So shaded artifacts can be used as libraries. See https://github.com/sbt/sbt-assembly/issues/350
This adds an extra ScalaSigProcessor step to the current jarjar steps. I copied MainProcessor into JJProcessor and scalaified it to be able to add the extra step.

This is still a WIP. We plan to work on the in our next sprint. I made this PR to gather feedback
on the approach.

There are some things that need to be done:
- Support @ScalaLongSignature|
- Support .sig files
- Add tests (scripted + scalatest)
- Idea: limitation on the rename rules that are supported. For example a new rule: renamePackage
    Reason it i find it difficult to oversee which errors may get introduced by supporting the rename   rules as they are right now.
- Support multiple scala versions. I currently only tested 2.12. Maybe we cannot use PickleBuffer the way we do now.
- Cleanup, split ScalaSigAnnotationProcessor into multiple class files.

I could really use some insight of a scala compiler developer to check if I've missed some obvious tags to rename.

### CODE REVIEW

File: src/main/scala/org/pantsbuild/jarjar/ScalaSigProcessor.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.pantsbuild.jarjar
+
+import org.objectweb.asm.{ClassReader, ClassWriter}
+import org.pantsbuild.jarjar.util.{EntryStruct, JarProcessor}
+
+class ScalaSigProcessor(rules: Seq[Rule]) extends JarProcessor {
+  override def process(struct: EntryStruct): Boolean = {
+    if (!struct.name.endsWith(".class") || struct.skipTransform) true
```
Comment (eed3si9n) [inline src/main/scala/org/pantsbuild/jarjar/ScalaSigProcessor.scala]: Maybe allow `*.sig` files here? - https://github.com/scala/scala/pull/7712

**Comment by eed3si9n** [comment]
Assuming shading is needed by tools other than sbt-assembly, I wonder if it might be useful to split this as a library on its own. When it starts supporting library, there wouldn't be a need to shade only at the end of fat JAR? If so shading could potentially stand on its own as a plugin (we can continue to provide integration from sbt-assembly since it seems to be a popular feature), or even a CLI tool.

**Comment by eed3si9n** [comment]
How about calling it Jar Jar Abrams?

**Comment by jeroentervoorde** [comment]
@eed3si9n About creating a new library (a can't reply to your comment, maybe we should move discussion to the issue?)

We have only limited time to work on this so i was hoping to continue this as a part of sbt-assembly for now if you don't mind. We'll keep in mind that we want to extract in the future, once it gets some real life testing.

You're right we are not limited to fat jars anymore when this is done. I used sbt-assembly to publish a shaded, fat, library in a test project to test this so currently it already fits that use case. I do like the idea of an sbt-shade plugin but i don't think we have time to build that now (and we do like to start using the shaded jars internally asap)

Something else:

I did run into an issue with ivy though which does not elide all the shaded dependencies. I actually don't know if ivy dependencies can be excluded for an additional artifact with classifier. Do you?

File: src/main/scala/org/pantsbuild/jarjar/ScalaSigProcessor.scala (hunk 1 above)
Comment (jeroentervoorde) [reply src/main/scala/org/pantsbuild/jarjar/ScalaSigProcessor.scala]: I'll add it to the list. I have some reading up to do here.

**Comment by eed3si9n** [comment]
> I did run into an issue with ivy though which does not elide all the shaded dependencies. I actually don't know if ivy dependencies can be excluded for an additional artifact with classifier. Do you?

I have a FAQ entry here that might be relevant - https://github.com/sbt/sbt-assembly#q-despite-the-concerned-friends-i-still-want-publish-fat-jars-what-advice-do-you-have

**Comment by jeroentervoorde** [comment]
Hey @eed3si9n , can you take a look at the PR?

I implemented most of the things from the initial todo list. I skipped .sig file support as i am not really sure this is actually useful in this plugin (and also how i'd test it)

**Comment by eed3si9n** [comment]
Also I'm assuming you consent to releasing with the licensing terms of sbt-assembly? (MIT License)
I am asking this here since as noted above we might want to pull apart shading feature as a stand-alone library.

**Comment by jeroentervoorde** [comment]
Sure! I consent to the licensing terms. Thanks for the review!

**Comment by eed3si9n** [comment]
I've split this into a new library - http://eed3si9n.com/jarjar-abrams

**Comment by jeroentervoorde** [comment]
@eed3si9n 

Cool! Are you willing to add my company name (Simacan) to the blog post as I got to do most of the work in office hours?

**Comment by eed3si9n** [comment]
Absolutely. I just added a line.

**Pull request merged**